### PR TITLE
Refactor handles field, add EntitySpec.deserialize for PA

### PR DESCRIPTION
**Pull request by cromwellian**
Adds EntitySpec.deserialize(map)
Adds HandleHolder interface which requires 'map' and 'entitySpec' fields
Adds handles field to Particle (HandleHolder)
HandleHolder delegates each handle field by map
Allows ArcHost to dynamically lookup and supply handles and specs

### CODE REVIEW

File: java/arcs/sdk/Entity.kt [hunk 1]
```diff
@@ -24,4 +24,12 @@ interface Entity {
 interface EntitySpec<T : Entity> {
     /** Returns an empty new instance of [T]. */
     fun create(): T
+
+    /**
+     * Takes a [Map] representing a serialized representation of a [RawEntity] from the
+     * storage layer, and converts it to a concrete entity class.
+     * TODO: replace this with kotlinx.serialization
+     * TODO: this could just accept a RawEntity, but it would be tighter-coupling of SDK<->Core
```
Comment (csilvestrini) [inline java/arcs/sdk/Entity.kt]: I think we decided it's ok for sdk to depend on core, so this might as well be a RawEntity. The Entity class already has a toRawEntity() method.

It's possible I misunderstood the outcome of that discussion though.

File: java/arcs/sdk/Particle.kt [hunk 2]
```diff
@@ -11,8 +11,33 @@
 
 package arcs.sdk
 
+/**
+ * Interface used by [ArcHost]s to interact dynamically with code-generated [Handle] fields
+ * used by [Particle]s.
+ */
+interface HandleHolder {
+    val map: MutableMap<String, Handle>
```
Comment (csilvestrini) [inline java/arcs/sdk/Particle.kt]: Nit: say something about what goes in the map. Either `handleMap` or just `handles` would be fine.

File: java/arcs/sdk/Particle.kt [hunk 3]
```diff
@@ -11,8 +11,33 @@
...
+    val entitySpecs: MutableMap<String, EntitySpec<out Entity>>
```
Comment (csilvestrini) [inline java/arcs/sdk/Particle.kt]: These don't need to be mutable, if you follow some of my other suggestions in this file.

File: java/arcs/sdk/Particle.kt [hunk 4]
```diff
@@ -11,8 +11,33 @@
...
+}
+
+/**
+ * Base class used by `schema2kotlin` code-generator tool to generate a class containing all
+ * declared handles.
+ */
+abstract class HandleHolderBase(
```
Comment (csilvestrini) [inline java/arcs/sdk/Particle.kt]: I don't think you need this class at all. The code generator can generate code that looks like this:

```
class FooHandles : HandleHolder {
    override val map = mapOf(
        "foo" to SingletonImpl(...),
        ...
    )
}
```

Then you don't need the empty mutable map to already exist.

File: java/arcs/sdk/Particle.kt [hunk 5]
```diff
@@ -11,8 +11,33 @@
...
+    override val map: MutableMap<String, Handle> = mutableMapOf(),
+    override val entitySpecs: MutableMap<String, EntitySpec<out Entity>> = mutableMapOf()
+) : HandleHolder
+
 /** Base interface for all particles. */
 interface Particle {
+
+    /**
+     * This field is provided in an abstract base class by the `schema2kotlin` code generator
+     * and contains all declared handles in the manifest.
```
Comment (csilvestrini) [inline java/arcs/sdk/Particle.kt]: This is probably too much implementation detail for a particle author to read.

I would reword the comment to be more about why a particle author would want to look at it (e.g. "contains all of the particle's handles" or something like that)

File: src/tools/schema2kotlin.ts [hunk 6]
```diff
@@ -83,15 +85,21 @@ ${this.opts.wasm ? 'import arcs.sdk.wasm.*' : 'import arcs.core.data.RawEntity\n
         default:
           throw new Error(`Unsupported handle direction: ${connection.direction}`);
       }
-      handleDecls.push(`val ${handleName}: ${handleInterfaceType} = ${this.getType(handleConcreteType) + 'Impl'}(particle, "${handleName}", ${entityType}_Spec())`);
+      handleDecls.push(`val ${handleName}: ${handleInterfaceType} by map`);
+      specDecls.push(`entitySpecs["${handleName}"] = ${entityType}_Spec()`);
+
+      // handleDecls.push(`val ${handleName}: ${handleInterfaceType} = ${this.getType(handleConcreteType) + 'Impl'}(particle, "${handleName}", ${entityType}_Spec())`);
```
Comment (csilvestrini) [inline src/tools/schema2kotlin.ts]: Delete?

File: src/tools/schema2kotlin.ts [hunk 7]
```diff
@@ -243,7 +255,7 @@ ${this.opts.wasm ? `
         )
         _rtn.internalId = internalId
         return _rtn
-    }` : ''}
+    }` : ''}
```
Comment (csilvestrini) [inline src/tools/schema2kotlin.ts]: Revert

File: java/arcs/sdk/Particle.kt (hunk 4 above)
Comment (cromwellian) [reply java/arcs/sdk/Particle.kt]: The code generator isn't allowed to instantiate these Handles, they must be instantiated by ArcHost via HandleManager, which is platform dependent.  Additionally, the StorageKeys are created by the Allocator, which doesn't happen until runtime. 

At best, these could be some kind of placeholder flyweights which are lazily initialized in place with real handles, but some kind of mutation is unavoidable.

Comment (cromwellian) [reply java/arcs/sdk/Particle.kt]: I think we can fix this in a different way by code-gening a method to create the handle classes and take the handle map and initialize the handles field, e.g.

abstract class BaseParticle<T> : Particle {
   <T : HandleHolder> setupHandles(map: Map<String, Handle>)
}

And then have the AbstractFooParticle<FooHandles> class code-gen this method. 

Then ArcHost can create all the handles, create an immutable map, and invoke BaseParticle.setupHandles(map)

Comment (cromwellian) [reply java/arcs/sdk/Particle.kt]: Making these truly immutable turned out to be more trouble than it was worth, so I instead simply declared them as Map<> but behind the scenes a MutableMap is used, which ArcHost will cast to. 

The only real way to achieve real immutability would be to make Particle ctor take these params, but we want to keep particle instantiation as arg-less ctors.

File: src/tools/schema2kotlin.ts (hunk 6 above)
Comment (cromwellian) [reply src/tools/schema2kotlin.ts]: branched (wasm vs jvm)

File: java/arcs/sdk/Particle.kt (hunk 4 above)
Comment (csilvestrini) [reply java/arcs/sdk/Particle.kt]: Fair enough, thanks for investigating
